gitpod-io/gitpod
# Add "has-maturity-level" admission constraint

## Description
This PR introduces a new admission constraint: `has-maturity-level`. There are three maturity levels: `low`, `default`, `high`.

In combination with a new `mature-workspace-clusters` permission we can now mark risk-averse users. Once a user has this permission, they will prefer clusters with the `high` maturity level. If no such cluster exists, they'll fall back to the usual cluster selection. This is implemented just by adding a new workspace cluster set to the cluster selection.

This PR is part of a stack of changes:
- https://github.com/gitpod-io/gitpod/pull/11383
- https://github.com/gitpod-io/gitpod/pull/11384
- ➡️ https://github.com/gitpod-io/gitpod/pull/11385

## How to test
1. start a workspace - nothing should change for regular users
2. add the `mature-workspace-clusters` permission to your user using the `admin` panel. Starting a workspace should work as it did before.
3. register a new cluster (e.g. by copying and modifying the entry in the database) and add the `has-maturity-level` admission constraint. Your user should now land on that cluster.

## Release Notes
```release-notes
NONE
```

## Werft options:
